src/python/WMCore/Services/Service.py

```diff
@@ -294,6 +298,10 @@ def getData(self, cachefile, url, inputdata=None, incoming_headers=None,
                     cachefile.write(data)
                     cachefile.seek(0, 0)  # return to beginning of file
                 else:
+                    if binary:
+                        with open(cachefile, 'wb') as f:
+                            f.write(data)
+                        return
```

Comment:
Not much. It's just that I find `return` statements in nested structures not to be very readable.
IMO, this code would like cleaner and more readable if written like:
```
            else:
                # Don't need to prepend the cachepath, the methods calling
                # getData have done that for us
                if isfile(cachefile):
                    cachefile.write(data)
                    cachefile.seek(0, 0)  # return to beginning of file
                elif binary:
                    with open(cachefile, 'wb') as f:
                        f.write(data)
                else:
                    with open(cachefile, 'w', encoding='utf-8') as f:
                        if isinstance(data, dict) or isinstance(data, list):
                            f.write(json.dumps(data))
                        else:
                            f.write(data)
```

or even a better version would be:
```
```
            else:
                # Don't need to prepend the cachepath, the methods calling
                # getData have done that for us
                if isfile(cachefile):
                    cachefile.write(data)
                    cachefile.seek(0, 0)  # return to beginning of file
                elif binary:
                    with open(cachefile, 'wb') as f:
                        f.write(data)
                elif isinstance(data, (dict, list))):
                    with open(cachefile, 'w', encoding='utf-8') as f:
                        f.write(json.dumps(data))
                else:
                    with open(cachefile, 'w', encoding='utf-8') as f:
                        f.write(data)
```

Having these flat blocks of if/elif/else look much better than nesting them, IMO. If you disagree, feel free to keep your implementation though, after all "being readable" can be subjective.

Reply:
I think the if/elif/else is wrong since if cachefile is given we need to seek to beginning of that file and then the data is written over there. I did not implement original logic there. So, after reading the code I think it should be like that:
```
                if binary:
                    with open(cachefile, 'wb') as f:
                        f.write(data)
                    return
                if isfile(cachefile):
                    cachefile.seek(0, 0)  # return to beginning of file
                with open(cachefile, 'w', encoding='utf-8') as f:
                    if isinstance(data, dict) or isinstance(data, list):
                        f.write(json.dumps(data))
                    else:
                        f.write(data)
```

Please review the logic of using cachefile and let me know. I think binary mode is independent from cachefile and it can come first since it writes into cachefile in binary mode. Then, if cache file exists we can seek to its beginning and write the data accordingly.

Reply:
I don't understand the whole logic here, but I think `cachefile` is actually a file descriptor which is somehow accessed by the subclass to read the content, hence seeking back to the beginning of the file.

The logic you have implemented in the code looks correct to me, I was just trying to make those nested if/else look more readable. Code wise, I think both of my suggestions are correct and do not change any logic in the code, just a different implementation.

Reply:
Alan, you lost me. The logic in the code please see file itself is differ from logic I just put in this comment. Please reply which one you think is correct:

### existing logic
```
                if isfile(cachefile):
                    cachefile.write(data)
                    cachefile.seek(0, 0)  # return to beginning of file
                else:
                    if binary:
                        with open(cachefile, 'wb') as f:
                            f.write(data)
                        return
                    with open(cachefile, 'w', encoding='utf-8') as f:
                        if isinstance(data, dict) or isinstance(data, list):
                            f.write(json.dumps(data))
                        else:
                            f.write(data)
```
As you can see, if cachefile is given the data will be written first to cachefile otherwise we apply different logic to different modes (binary or not).

### proposed logic version
```
                # if cachefile is a file description we write data to it and seek at the beginning of the file
                if isfile(cachefile):
                    cachefile.write(data)
                    cachefile.seek(0, 0)  # return to beginning of file
                    return

                # if binary mode is given we write data in binary mode
                if binary:
                    with open(cachefile, 'wb') as f:
                        f.write(data)
                    return

                # if cachefile is a file name but not in binary mode we open it and write data accordingly
                with open(cachefile, 'w', encoding='utf-8') as f:
                    if isinstance(data, dict) or isinstance(data, list):
                        f.write(json.dumps(data))
                    else:
                        f.write(data)
```
Please note, here I return early when we hit individual use-case and logic does not need nested if/else/elif part.

Please say, and write, explicitly which logic I should keep.

Reply:
Note that if/elif/else is different than a case statement, in the sense that whenever one of the if statements evaluate to True, all the other statements are skipped. Having said that, I would replace the existing logic (which you also pasted above in your last comment), by the following (which I also mentioned in one of my previous replies):
```
        if from_cache:
            # If it's coming from the cache we don't need to write it to the
            # second cache, or do we?
            self['logger'].debug('Data is from the Requests cache')
        else:
            # Don't need to prepend the cachepath, the methods calling
            # getData have done that for us
            if isfile(cachefile):
                cachefile.write(data)
                cachefile.seek(0, 0)  # return to beginning of file
            elif binary:
                with open(cachefile, 'wb') as f:
                    f.write(data)
            elif isinstance(data, (dict, list))):
                with open(cachefile, 'w', encoding='utf-8') as f:
                    f.write(json.dumps(data))
            else:
                with open(cachefile, 'w', encoding='utf-8') as f:
                    f.write(data)
```

which has the same logic as your proposal above, but it avoids using `return` statements in favor of an if/elif/else logic.

Reply:
done